src/editors/EditorHeader.jsx

```diff
@@ -1,44 +1,112 @@
 import {
-  ActionRow, IconButton, Icon, ModalDialog,
+  ActionRow, IconButton, Icon, ModalDialog, Form,
 } from '@edx/paragon';
-import PropTypes from 'prop-types';
-import React, { useContext } from 'react';
-import { Close } from '@edx/paragon/icons';
+import React, { useContext, useState, useEffect } from 'react';
+import { Edit, Close } from '@edx/paragon/icons';
 import EditorPageContext from './EditorPageContext';
-import { ActionStates, mapBlockTypeToName } from './data/constants';
+import { ActionStates } from './data/constants';
 
-const EditorHeader = ({ title }) => {
-  const { unitUrl, unitUrlLoading, studioEndpointUrl } = useContext(EditorPageContext);
+/* The Header component renders based on three states:
+    loading -> loaded but not editing -> editing. */
+
+const EditorHeader = () => {
+  const {
+    titleRef, editorRef, setBlockTitle,
+    blockValue, blockLoading,
+    unitUrl, unitUrlLoading, studioEndpointUrl,
+  } = useContext(EditorPageContext);
+
+  const [editing, setEditing] = useState(false);
+
+  const [title, setTitle] = useState('');
+  useEffect(() => {
+    if (blockLoading === ActionStates.FINISHED) {
+      setTitle(blockValue ? blockValue.data.display_name : '');
+    }
+  }, [blockLoading]);
+
+  const handleChange = (e) => {
+    setTitle(e.target.value);
+  };
+
+  const handleKeyDown = (e) => {
+    if (e.key === 'Enter') { setEditing(false); }
+    if (e.key === 'Tab' && editorRef) {
+      e.preventDefault();
+      editorRef.current.focus();
+    }
+  };
+
+  const handleOnBlur = () => {
+    setBlockTitle(title);
+    setEditing(false);
+  };
 
   const onCancelClicked = () => {
     if (unitUrlLoading === ActionStates.FINISHED) {
       const destination = `${studioEndpointUrl}/container/${unitUrl.data.ancestors[0].id}`;
       window.location.assign(destination);
     }
   };
+
+  let titleDisplay = 'Loading...';
+  if (blockLoading === ActionStates.FINISHED) {
+    if (!editing) {
+      titleDisplay = (
+        <div lassName="d-flex">
+          <div style={{ lineHeight: '1.5', paddingRight: '.25em' }}>
+            {title || 'Title'}
+          </div>
+          <IconButton
+            src={Edit}
+            iconAs={Icon}
+            aria-label="Edit Title"
+            alt="Edit"
+            onClick={() => setEditing(true)}
+            className="mr-2"
+            size="sm"
+            autoFocus
+          />
+        </div>
+      );
+    } else {
+      titleDisplay = (
+        <Form.Group>
+          <Form.Control
+            ref={(input) => { titleRef.current = input; }}
+            value={title || ''}
+            placeholder="Title"
+            onChange={handleChange}
+            onKeyDown={handleKeyDown}
+            onBlur={handleOnBlur}
+            autoFocus
+            trailingElement={<Icon src={Edit} />}
+          />
+        </Form.Group>
+      );
+    }
+  }
```

Comment:
The key logic for what is displayed in the title.
When the block is loading, this area should only show 'Loading...'.
When the user is not editing, this area will display the text with an edit button following it.
When the user is editing, this area will display a text input.